hanish/darksky.py

```diff
@@ -0,0 +1,192 @@
+# hanish.darksky
+# Handlers and functionality for making Dark Sky API requests.
+#
+# Author:   Benjamin Bengfort <benjamin@bengfort.com>
+# Created:  Mon May 08 12:04:11 2017 -0400
+#
+# Copyright (C) 2016 Bengfort.com
+# For license information, see LICENSE.txt
+#
+# ID: cdarksky.py [] benjamin@bengfort.com $
+
+"""
+Handlers and functionality for making Dark Sky API requests.
+"""
+
+##########################################################################
+## Imports and Module Vars
+##########################################################################
+
+import requests
+
+try:
+    # Python 3
+    from urllib.parse import urljoin
+except ImportError:
+    # Python 2
+    from urlparse import urljoin
+
+
+from datetime import date, datetime
+from expiringdict import ExpiringDict
+from .exceptions import HanishValueError
+from .exceptions import DarkSkyException
+
+
+DARKSKY_API_URL  = "https://api.darksky.net"
+API_CALLS_HEADER = "X-Forecast-API-Calls"
+
+
+##########################################################################
+## DarkSky API
+##########################################################################
+
+class DarkSky(object):
```

Comment:
That's an excellent question, and one I hoped that might be asked when I included it here. I'm a fan of microservice architectures because I believe they make it possible to iterate on large systems with lots of components. In order to create production systems that use microservices, minimizing latency is crucial, and so caching must be a first thought, not a bolt-on after design. 

In terms of options, there are quite a few, and for me it depends on what I'm caching and where. What comes to mind is in-process caching for single and multiple threads, http caching, and general purpose caching. Another crucial requirement is if the cache is volatile or needs to be durable. 

For inprocess caching with a single thread, it's hard to beat something like the `ExpiringDict`. If the cache was going to grow to large sizes, I would probably add a garbage collection routine that locks the cache and removes expired values, the current implementation only removes expired values on access. 

Moving to multiple threads, I would put the cache in it's own thread and surround it with read and write locks for consistency (since most of the time the cache is being read, this would still be very available). I would also probably use an in-memory sqlite database with a key index as the data store. However, if durability was required, I would consider an embedded key-value store like BoltDB or LevelDB, both of which are very fast, have checkpointing and also have local in-memory caching. 

For caching between processes or services, initially I would use simple HTTP caches, and it's hard to beat Memcached or a cloud provider's implementation like ElastiCache on AWS. It's fairly easy to tune and configure for durability and load balancing, so it's a natural first step. 

For general purpose caching, both for single node and distributed, I'd move towards Redis or MemSQL depending on the complexity of the data being cached. I know that Redis is being used more and more for HTTP caching, but I've used it as a local data store routinely. I think it's harder to shard/distribute these types of services, but it can be worth it when you want to add logic to how caches are maintained and managed. 

Ok, this is a long answer, but by no means complete! Briefly I would mention message queues (Kafka, RabbitMQ, SQS) as one option for caching depending on use case. Also data stores like Riak and MongoDB could also be adapted as caches for long-term caching, directly using the REST APIs that they provide.